### PR TITLE
Filter on Gyms

### PR DESCRIPTION
**Pull request by alp-bitcoin**
Provide a way to filter gyms.

## Motivation and Context
Identifies gyms that do not need a fight to add pokemon, sort by team and last scan.

Open gyms are gyms with a higher level than number of Pokemon in it.

Team filter allows any team, or any particular team.

Last scan has options of "last hour, last 6 hours, last 24 hours, last week".  This is useful if you need to make sure the gym has been recently scanned and don't waste time going to a gym that has changed hands or has been filled up long since your scan.

## Screenshot

gymfilter

### CODE REVIEW

**Comment by tr1p0z** [comment]
How can I check this in git? Can i download this with git?

**Comment by alp-bitcoin** [comment]
@tr1p0z - You clone my fork and use it if you'd like to test it or use it.  I plan a few more enhancements to it and would welcome any feedback.

**Comment by tr1p0z** [comment]
For me is not working, I dont see gyms

**Comment by tr1p0z** [comment]
You sure it's showing all gyms with space? I'm scanning 1 hour ago and found 0 and my bot is getting gyms

**Comment by alp-bitcoin** [comment]
@tr1p0z  - You need to be scanning with the -gi option for this to work.  Otherwise it won't know how many pokemon are in the gym.  If you hover over a gym, you can see if it's open by seeing that it has a level higher than the number of Pokemon visible (e.g. it will say Level 3, 4000/8000, and have only 2 guys in it).

**Comment by alp-bitcoin** [comment]
This has been enhanced to also include gym level filters.  I plan to add more options for open spots to allow finding gyms that are within X prestige of leveling up.

**Comment by alp-bitcoin** [comment]
Added a "near prestige" filter.  I plan on making no more feature additions unless problems are found or feedback is needed to be addressed.

**Comment by jabberjabber** [comment]
I might have merged incorrectly but when I select "<1000 prestige until open spot" it also shows level 10 gyms, which it shouldn't

**Comment by Shentang** [comment]
Maximum level value should be 10 by default - otherwise there are no gyms shown.

File: pogom/utils.py [hunk 1]
```diff
@@ -266,7 +266,7 @@ def get_args():
                     if field_error != '':
                         type_error = 'empty!'
                         if field_error == 'method':
-                            type_error = 'not ptc or gmail instead we got \'' + fields[0] + '\'!'
+                            type_error = 'not ptc or google instead we got \'' + fields[0] + '\'!'
```
Comment (jnovack) [inline pogom/utils.py]: The changes here have nothing to do with this PR.  Please revert this file and resubmit this file only as a new PR.

File: static/data/pokemon.json [hunk 2]
```diff
@@ -1811,7 +1811,7 @@
   },
   "143": {
     "name": "Snorlax",
-    "rarity": "Rare",
+    "rarity": "Very Rare",
```
Comment (jnovack) [inline static/data/pokemon.json]: This was changed already in a previous PR.  Please rebase.

File: templates/map.html [hunk 3]
```diff
@@ -65,6 +65,70 @@
...
+                  <option value="6">6</option>
+                  <option value="7">7</option>
+                  <option value="8">8</option>
+                  <option value="9">9</option>
+                  <option value="10">10</option>
+                </select>
+              </div>
+               <div class="form-control switch-container" id="last-update-gyms-wrapper">
+                 <h3>Last Scan</h3>
+                <select name="last-update-gyms-switch" id="last-update-gyms-switch">
```
Comment (jnovack) [inline templates/map.html]: Unfortunately, this only updates when prestige was lowered or raised, and not when the team captured control.

File: pogom/utils.py (hunk 1 above)
Comment (alp-bitcoin) [inline pogom/utils.py]: Will revert.  Not sure how this got changed.

File: templates/map.html (hunk 3 above)
Comment (alp-bitcoin) [inline templates/map.html]: I'm not sure I follow on this one.

**Comment by alp-bitcoin** [comment]
Appreciate all the feedback.  I may move the min/max to be a multi-select list instead.

@jabberjabber - I thought I had a default value of 10, but will fix, or adjust to a mulit-select.

@Zerotask - nothing should have changed as part of this PR, so I'll make sure I'm rebased.

@Shentang - Good catch on the edge case of level 10!

File: pogom/utils.py (hunk 1 above)
Comment (alp-bitcoin) [inline pogom/utils.py]: I'm not seeing this change anymore.

File: static/data/pokemon.json (hunk 2 above)
Comment (alp-bitcoin) [inline static/data/pokemon.json]: I'm not seeing this file as changed anymore.

**Comment by alp-bitcoin** [comment]
Level 10 issue is now fixed.  Also fixed default max gym to 10.

File: templates/map.html (hunk 3 above)
Comment (jnovack) [inline templates/map.html]: There is no value in the database that stores when the gym SWITCHES teams.  There's only a value in database when the scanner last notices that something is different (including Prestige going higher or lower).

Comment (alp-bitcoin) [inline templates/map.html]: I'm not sure why that's relevant to this line of code.  The line of code you list here only refers to the last time it was actually scanned, and is html.  Nothing in this feature is tracking when a gym is changing prestige or team.

**Comment by jabberjabber** [comment]
I suggest changing last scan to be an input field (that accepts integers) followed by a drop down to select minutes, hours, days. 
Set input field to 0 to show all.

Same applies for open spot. Let us input our own number for prestige until open spot.

**Comment by alp-bitcoin** [comment]
Rebase completed.

**Comment by alp-bitcoin** [comment]
@jabberjabber I'm not sure that level of precision is all that useful compared to the effort, nor would it necessarily be a usability improvement.  I'd likely prefer a system with reasonable defaults, with a way to have a custom value.  I'll consider that for a future PR.  Same goes with the prestige - Selecting a few reasonable values from a dropdown is a lot more usable than fine-tuning exact prestige values.  Will definitely consider a "Custom" value which then displays a free-form input, but I'm working on other features that will provide more value, like scan optimization for gyms.

File: templates/map.html (hunk 3 above)
Comment (alp-bitcoin) [inline templates/map.html]: I think I understand this now.  The "switch" refers to a switch control, not that the gym switches.  The intent is to find when a gym was last scanned by the scanner, to determine how fresh the data was.  I don't want to go to a gym that I scanned 2 days ago, and find out that the open spot I saw is long gone.

**Comment by jabberjabber** [comment]
@alp-bitcoin sounds good.

**Comment by beccasafan** [comment]
If scanning with --gym-info is required for some of these filters, then they should probably be hidden when not applicable. (Just like some of the location search controls are hidden when scanning with --fixed-location.

**Comment by jnovack** [comment]
>  If scanning with --gym-info is required for some of these filters, then they should probably be hidden when not applicable. 

That's impossible to know.  The scanners do NOT have to be running in the same thread (or on the same box) as the webserver.

**Comment by alp-bitcoin** [comment]
@beccasafan There is only one part of this that doesn't work perfectly without -gi running, and as @jnovack mentioned, there's zero way to even know whether or not a scanner is running actively.

The only part that does not work is Open Spot with a value of Open Spots.  In that case, the behavior is still reasonable - it cannot determine anything with an open spot, so it shows nothing.  The "nearly next level prestige" values work just fine without -gi, as do all other filters (level, last scan, team).

You can remove the pending changes label, there are no pending changes, barring any bugs reported.

**Comment by alp-bitcoin** [comment]
There are no further updates, if this just needs a rebase, will be happy to do so.

**Comment by alp-bitcoin** [comment]
I'm removing this pull request since it seems unlikely to be merged.

**Comment by jabberjabber** [comment]
Why is this unlikely to be merged? I think the owners of this project just take a little time to work through things. Please re-open this, I would love to have this functionality.

**Comment by alp-bitcoin** [comment]
I got tired of pointless rebases for something unlikely to get merged, especially after it got tagged pending changes when there weren't any, and  the owners refused to remove it.

Pointless now since the API is dead anyway.

**Comment by gabaod** [comment]
Well I put this in recently and have noticed an issue when viewing the map via android phone.  All i see is the options settings that i can scroll up and down and thats it. totally overlays the map.   Works great via desktop browser.

**Comment by jabberjabber** [comment]
@alp-bitcoin You sure you don't want to keep this going? Looks like the project maintainers are a little more active these days. I really like this PR and would like to see it merged. If you re-open this I can help by ping-ing the devs on discord and push for this change.

**Comment by gabaod** [comment]
I agree.. great PR here.  One thing though is due to 1422 being merged in dev, it uses the same variable name for the wrapper thus this will need to be updated, plus when I used it it seems to only remove the Open gym field when I remove gyms, and just need to fix that so all of these extra details that we can sort by are hidden from options if gyms are not selected

**Pull request closed without merging**